sdks/python/apache_beam/io/gcp/spannerio.py

```diff
@@ -0,0 +1,559 @@
+               read_timestamp=None, exact_staleness=None, credentials=None,
+               sql=None, params=None, param_types=None,  # with_query
+               table=None, columns=None, index="", keyset=None,  # with_table
+               read_operations=None,  # for read all
+               transaction=None
+              ):
+    """
+    A PTransform that uses Spanner Batch API to perform reads.
+
+    Args:
+      project_id: Cloud spanner project id. Be sure to use the Project ID,
+        not the Project Number.
+      instance_id: Cloud spanner instance id.
+      database_id: Cloud spanner database id.
+      pool: (optional) Session pool to be used by database. If not passed, the
+        database will construct an instance of BurstyPool.
+      read_timestamp: (optional) An instance of the `datetime.datetime` object
+        to execute all reads at the given timestamp.
+      exact_staleness: (optional) An instance of `datetime.timedelta` object
+        to execute all reads at a timestamp that is exact_staleness old.
+      credentials: (optional) The authorization credentials to attach to
+        requests. These credentials identify this application to the service.
+        If none are specified, the client will attempt to ascertain
+        the credentials from the environment.
+      sql: (optional) SQL query statement
+      params: (optional) Values for parameter replacement. Keys must match the
+        names used in sql.
+      param_types: (optional) maps explicit types for one or more param values;
+        required if params are passed.
+      table: (optional) Name of the table from which to fetch data.
+      columns: (optional) List of names of columns to be retrieved; required if
+        the table is passed.
+      index: (optional) name of index to use, rather than the table's primary
+        key.
+      keyset: (optional) keys / ranges identifying rows to be retrieved.
+      read_operations: (optional) List of the objects of :class:`ReadOperation`
+        to perform read all.
+      transaction: (optional) PTransform of the :meth:`create_transaction` to
+        perform naive read on cloud spanner.
+    """
+    self._configuration = _BeamSpannerConfiguration(
+        project=project_id, instance=instance_id, database=database_id,
+        credentials=credentials, pool=pool,
+        snapshot_read_timestamp=read_timestamp,
+        snapshot_exact_staleness=exact_staleness
+    )
+
+    self._read_operations = read_operations
+    self._transaction = transaction
+
+    if self._read_operations is None:
+      if table is not None:
+        if columns is None:
+          raise ValueError("Columns are required with the table name.")
+        self._read_operations = [ReadOperation.table(
+            table=table, columns=columns, index=index, keyset=keyset)]
+      elif sql is not None:
+        self._read_operations = [ReadOperation.query(
+            sql=sql, params=params, param_types=param_types)]
+
```

Comment:
Don't we need an error here if both table and sql are empty?

Reply:
I implemented the checks in the expand method. https://github.com/apache/beam/pull/9606/files#diff-fbff531869ead87113e7c97e085d7013R513